feat: add `app.onStart()` and `app.onStop()` helpers

bajtos opened the pull request:
In some cases,  it's desirable to register a single function to be called at start or stop time. For example, when writing integration-level tests, we can use `app.onStop()` to register a cleanup routine to be invoked whenever the application is shut down.

This pull request adds `app.onStart()` and `app.onStop()` methods that can be used to register life-cycle observers as plain functions.

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- Affected artifact templates in `packages/cli` were updated
- Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

raymondfeng commented on packages/core/src/application.ts:
```diff
@@ -312,6 +314,23 @@ export class Application extends Context implements LifeCycleObserver {
+   * @returns The LifeCycleObserver binding created.
+   */
+  public onStart(fn: () => ValueOrPromise<void>): Binding<LifeCycleObserver> {
+    const key = `life-cycle-observers.${fn.name || '<onStart>'}-${uuid()}`;
```
Let's use `CoreBindings.LIFE_CYCLE_OBSERVERS` as the namespace.

raymondfeng replied:
Let's use [BindingKey.generate](https://github.com/strongloop/loopback-next/blob/master/packages/context/src/binding-key.ts#L133) to generate the key.

bajtos replied:
> Let's use `CoreBindings.LIFE_CYCLE_OBSERVERS` as the namespace.

+1, will do. Thank you for pointing out this constant, I was not aware of it.



> Let's use [BindingKey.generate](https://github.com/strongloop/loopback-next/blob/master/packages/context/src/binding-key.ts#L133) to generate the key.

I want to keep function name in the binding key. `BindingKey.generate` does not support that :(

bajtos replied:
I briefly considered enhancing `BindingKey.generate` with a new optional argument `name`, but such change requires additional tests & docs to be written, which seems like too much overhead for what I actually need in this pull request.

raymondfeng commented on packages/core/src/application.ts:
```diff
@@ -335,6 +354,23 @@ export class Application extends Context implements LifeCycleObserver {
+   * @returns The LifeCycleObserver binding created.
+   */
+  public onStop(fn: () => ValueOrPromise<void>): Binding<LifeCycleObserver> {
+    const key = `life-cycle-observers.${fn.name || '<onStop>'}-${uuid()}`;
```
Ditto.

raymondfeng commented on packages/core/src/application.ts:
```diff
@@ -312,6 +314,23 @@ export class Application extends Context implements LifeCycleObserver {
+    const key = `life-cycle-observers.${fn.name || '<onStart>'}-${uuid()}`;
+    return this.bind<LifeCycleObserver>(key)
+      .to({start: fn})
+      .tag(CoreTags.LIFE_CYCLE_OBSERVER);
```
Let's use `.apply(asLifeCycleObserver)` for consistency. See https://github.com/strongloop/loopback-next/blob/master/packages/core/src/lifecycle.ts#L58

bajtos replied:
Of course! Thanks for pointing out this binding template function to me :+1:

raymondfeng commented on packages/core/src/application.ts:
```diff
@@ -312,6 +314,23 @@ export class Application extends Context implements LifeCycleObserver {
+   * or asynchronous (returning `Promise<void>`).
+   * @returns The LifeCycleObserver binding created.
+   */
+  public onStart(fn: () => ValueOrPromise<void>): Binding<LifeCycleObserver> {
```
Do we want to allow an optional `group` as the 2nd argument for ordering the life cycle observer?

bajtos replied:
It makes sense to me to eventually support `group` argument, but since I don't need it myself, I prefer to leave that out of my pull request. [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it).

raymondfeng replied:
It would be nice to support it as part of this PR to make the feature complete without requiring us to revisit it. But feel free to defer it.

bajtos replied:
> It would be nice to support it as part of this PR to make the feature complete without requiring us to revisit it.

I know! On the other hand, if we leave `group` out and somebody comes back to use asking for that feature, then we will validate usefulness of this feature and provide a good opportunity for the reporter to contribute the fix, growing our community of code contributors by one more person.

bajtos commented:
@raymondfeng Thank you for a great feedback. I addressed most of your comments in 8ad88d0, also added a tiny tsdoc improvement in be7b590.

LGTY now?
